hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -960,6 +968,119 @@ class Emit[C](
+        val sorter = new ArraySorter(EmitRegion(mb, region.code), sortedElts)
+
+        val (k1, k2) = keyValTyp match {
+          case t: PStruct => GetField(In(0, t), "key") -> GetField(In(1, t), "key")
+          case t: PTuple =>
+            assert(t.fields(0).index == 0)
+            GetTupleElement(In(0, t), 0) -> GetTupleElement(In(1, t), 0)
+        }
+
+        val compare = ApplyComparisonOp(Compare(keyValTyp.types(0).virtualType), k1, k2) < 0
+        InferPType(compare)
+        val leftRightComparatorNames = Array.empty[String]
+        val sortF = sortedElts.ti match {
+          case BooleanInfo => makeDependentSortingFunction[Boolean](region.code, keyValTyp, compare, env, leftRightComparatorNames)
+          case IntInfo => makeDependentSortingFunction[Int](region.code, keyValTyp, compare, env, leftRightComparatorNames)
+          case LongInfo => makeDependentSortingFunction[Long](region.code, keyValTyp, compare, env, leftRightComparatorNames)
+          case FloatInfo => makeDependentSortingFunction[Float](region.code, keyValTyp, compare, env, leftRightComparatorNames)
+          case DoubleInfo => makeDependentSortingFunction[Double](region.code, keyValTyp, compare, env, leftRightComparatorNames)
+        }
+
+        val groupSizes = new StagedArrayBuilder(PInt32(), mb, 0)
+
+        val (lastKey, currKey) = (keyValTyp.virtualType: @unchecked) match {
+          case ts: TStruct =>
+            GetField(In(0, keyValTyp), ts.fieldNames(0)) -> GetField(In(1, keyValTyp), ts.fieldNames(0))
+          case tt: TTuple =>
+            GetTupleElement(In(0, keyValTyp), tt.fields(0).index) -> GetTupleElement(In(1, keyValTyp), tt.fields(0).index)
+        }
+        val compare2 = ApplyComparisonOp(EQWithNA(keyTyp.virtualType), lastKey, currKey)
+        InferPType(compare2)
+        val isSame = mb.genEmitMethod("isSame",
+          FastIndexedSeq(typeInfo[Region], keyValTyp.asEmitParam, keyValTyp.asEmitParam),
+          BooleanInfo)
+        isSame.emitWithBuilder { cb =>
+          emitInMethod(cb, compare2).consumeCode[Boolean](cb, true, _.tcode[Boolean])
+        }
+
+        val xOut = mb.newPLocal("groupByKey_result", arrayTyp)
+        val eltIdx = mb.newLocal[Int]("groupByKey_eltIdx")
+        val grpIdx = mb.newLocal[Int]("groupByKey_grpIdx")
+        val withinGrpIdx = mb.newLocal[Int]("groupByKey_withinGrpIdx")
+        val outerSize = mb.newLocal[Int]("groupByKey_outerSize")
+        val groupSize = mb.newLocal[Int]("groupByKey_groupSize")
+
+        val outerRegion = region.asParent(collectionTyp.separateRegions, "GroupByKey")
+        emitStream(collection, outerRegion).map(cb) { stream =>
+          cb += EmitStream.write(mb, stream.asStream, sortedElts, outerRegion)
+          cb += sorter.sort(sortF)
+          cb += sorter.pruneMissing
+          cb.ifx(sortedElts.size.ceq(0), {
+            cb.assign(xOut, arrayTyp.constructEmpty(cb, region.code))
```

Comment:
can we take this out? We shouldn't need to special-case 0 here.

Reply:
I don't think the other branch handles the empty case as is, but I'm sure it's possible to tweak it so that it does. I will try.

Reply:
if, not can just call constructFromFunctions with 0 and finish